src/bmp/decoder.rs

```diff
@@ -109,17 +109,129 @@ impl<'a> Iterator for RowIterator<'a> {
+    /// Specified image type is invalid for top-down BMPs (i.e. is compressed)
+    ImageTypeInvalidForTopDown(u32),
+    /// Image type not currently recognized by the decoder
+    ImageTypeUnknown(u32),
+
+    /// Bitmap header smaller than the core header
+    HeaderTooSmall(u32),
+
+    /// The palette is bigger than allowed by the bit count of the BMP
+    PaletteSizeExceeded {
+        colors_used: u32,
+        bit_count: u16,
+    }
+}
+
+impl fmt::Display for BmpDecoderError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            BmpDecoderError::SizeOverflow =>
+                f.write_str("Image would require a buffer that is too large to be represented!"),
+            BmpDecoderError::RleDataTooShort =>
+                f.write_str("Not enough RLE data"),
+            BmpDecoderError::BitfieldMaskNonContiguous =>
+                f.write_str("Non-contiguous bitfield mask"),
+            BmpDecoderError::BitfieldMaskInvalid =>
+                f.write_str("Invalid bitfield mask"),
+            BmpDecoderError::BitfieldMaskMissing(bb) =>
+                f.write_fmt(format_args!("Missing {}-bit bitfield mask", bb)),
+            BmpDecoderError::BitfieldMasksMissing(bb) =>
+                f.write_fmt(format_args!("Missing {}-bit bitfield masks", bb)),
+            BmpDecoderError::BmpSignatureInvalid =>
+                f.write_str("BMP signature not found"),
+            BmpDecoderError::MoreThanOnePlane =>
+                f.write_str("More than one plane"),
+            BmpDecoderError::InvalidChannelWidth(tp, n) =>
+                f.write_fmt(format_args!("Invalid channel bit count for {}: {}", tp, n)),
+            BmpDecoderError::NegativeWidth(w) =>
+                f.write_fmt(format_args!("Negative width ({})", w)),
+            BmpDecoderError::ImageTooLarge(w, h) =>
+                f.write_fmt(format_args!("Image too large (one of ({}, {}) > soft limit of {})", w, h, MAX_WIDTH_HEIGHT)),
+            BmpDecoderError::InvalidHeight =>
+                f.write_str("Invalid height"),
+            BmpDecoderError::ImageTypeInvalidForTopDown(tp) =>
+                f.write_fmt(format_args!("Invalid image type {} for top-down image.", tp)),
+            BmpDecoderError::ImageTypeUnknown(tp) =>
+                f.write_fmt(format_args!("Unknown image compression type {}", tp)),
+            BmpDecoderError::HeaderTooSmall(s) =>
+                f.write_fmt(format_args!("Bitmap header too small ({} bytes)", s)),
+            BmpDecoderError::PaletteSizeExceeded { colors_used, bit_count } =>
+                f.write_fmt(format_args!("Palette size {} exceeds maximum size for BMP with bit count of {}", colors_used, bit_count)),
+        }
+    }
+}
+
+impl error::Error for BmpDecoderError {}
+
+/// Image types with distinct errors currently supported by `BmpDecoder`
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub enum BmpDecoderErrorInvalidChannelWidthImageType {
```

Comment:
Waaaaaay too long ;)

Reply:
`ChannelWidthError` essentially conveys the same information and is much shorter.

Reply:
Renamed to `DecoderErrorImageType`